src/app/treelist-checkbox/treelist-checkbox.component.ts

```diff
@@ -0,0 +1,250 @@
+        elements.forEach(element => {
+          (element as HTMLInputElement).checked = false;
+          (element as HTMLInputElement).indeterminate = false;
+        });
+        break;
+      }
+      case 1: {
+        const elements = document.querySelectorAll(`[id='${data.dataset.name}']`);
+        elements.forEach(element => {
+          (element as HTMLInputElement).checked = true;
+          (element as HTMLInputElement).indeterminate = true;
+        });
+        break;
+      }
+      case 2: {
+        const elements = document.querySelectorAll(`[id='${data.dataset.name}']`);
+        elements.forEach(element => {
+          (element as HTMLInputElement).checked = true;
+          (element as HTMLInputElement).indeterminate = false;
+        });
+        break;
+      }
+    }
+  }
+
+  public getAllChildren(
+    container: DatasetNode[], list: Set<string> = null, includeChilds: boolean = false
+  ): Set<string> {
+    if (list === null) {
+      list = new Set<string>();
+    }
+
+    container.forEach(child => {
+      if (child.children.length === 0) {
+        list = list.add(child.dataset.name);
+      } else {
+        if (includeChilds) {
+          list = list.add(child.dataset.name);
+        }
+        list = this.getAllChildren(child.children, list);
+      }
+    });
+    return list;
+  }
+
+  public getAllParents(json: any): Array<any> {
+    const attachParents = (inputNode: DatasetNode): any => {
+      inputNode.children.forEach(node => {
+        if (node.children.length > 0) {
+          node.children.forEach(child => {
+            child['parent'] = node;
+            attachParents(child);
+          });
+        }
+        node['parent'] = inputNode;
+      });
+      return inputNode;
+    };
+    const findNodeInTree = (tree_: DatasetNode, key: DatasetNode, result = []): Array<any> => {
+      if (tree_.dataset.name === key.dataset.name) {
+        result.push(tree_);
+      }
+      tree_.children?.forEach(node => {
+        if (node.dataset.name === key.dataset.name) {
+          result.push(node);
+        }
+        if (node.children) {
+          node.children.forEach(child => {
+            result = findNodeInTree(child, key, result);
+          });
+        }
+      });
+      return result as Array<DatasetNode>;
+    };
+    const children = findNodeInTree(
+      attachParents(this.data) as DatasetNode, this.findNodeById(this.data, json.dataset.name)
+    );
+    const temp = [];
+    children.forEach(parent => temp.push(parent.parent?.dataset?.id));
+    return children.map(parent => parent.parent?.dataset?.id);
+  }
+
+  public findNodeById(node: DatasetNode, id: string): DatasetNode | undefined {
+    if (node.dataset.name === id) {
+      return node;
+    }
+
+    if (node.children) {
+      for (const child of node.children) {
+        const foundNode = this.findNodeById(child, id);
+        if (foundNode) {
+          return foundNode;
+        }
+      }
+    }
+    return undefined;
+  }
```

Comment:
This method is identical to the `findNodeById` function that can be found in `study-filters-block.component.ts` on line `88`. Extract it to a single place if possible.